### PR TITLE
Support loading entrypoints by passing a module instead of a function

### LINKED ISSUE #58939: [FEATURE REQUEST] Salt's Entry Points should not need to be a callable
The way salt supports extensions is by means of entry-points, for example, on `setup.cfg`:
```ini
[options.entry_points]
salt.loader=
  runner_dirs = thirpartypackage.loader:func_to_get_list_of_dirs
  module_dirs = thirpartypackage.loader:func_to_get_list_of_dirs
```

This forces 3rd-party extensions to define a function which will return a list of paths from where the salt loader can then load modules.

Commonly, these 3rd-party extensions define all their loader related modules under a single package, to there's really no need for the entry point
to return a function, it can just return the module.
For example, on `setup.cfg`:
```ini
[options.entry_points]
salt.loader=
  runner_dirs = thirpartypackage.modules
  module_dirs = thirpartypackage.runners
```

This new approach reduces the amount of code to extend salt by a 3rd-party package.

### PR DESCRIPTION
**Pull request by s0undt3ch**
### What does this PR do?
Support loading entrypoints by passing a module instead of a function.

The old way of defining salt loader entrypoints was, for example, on `setup.cfg`:
```ini
[options.entry_points]
salt.loader=
  runner_dirs = thirpartypackage.loader:func_to_get_list_of_dirs
  module_dirs = thirpartypackage.loader:func_to_get_list_of_dirs
```
This old way is still supported in the rare ocasion where you actually
want to pass multiple paths to the loader and to not break backwards
compatibility.
    
The new way of defining the salt loader entrypoints is, for example, on `setup.cfg`:
```ini
[options.entry_points]
salt.loader=
  <this-name-does-not-matter = thirpartypackage
  <this-name-does-not-matter = thirpartypackage:callable
```
    
With this new approach, when the target is a package, Salt looks for the
`ext_type` being loaded as submodule of the package and if it exists, it
loads from there.
If the target is a callable, then, that callable should return a
dictionary where the keys are the several salt loader names. For
salt execution modules it would be `modules`, for salt state modules it would
be `states`, etc. The same directory layout you see in salt.
The values for those keys needs to be an iterable(not a string which is
also an iterable), ie, a list, a generator, etc, whose values are the
paths from where to load the salt loader modules.

### What issues does this PR fix or reference?

Fixes #58939

### CODE REVIEW

File: changelog/58939.added [hunk 1]
```diff
@@ -0,0 +1,23 @@
+Salt now supports loading entrypoints by passing a module instead of a function.
```
Comment (waynew) [inline changelog/58939.added:1]: This is probably excessive for the changelog. It will already link to the associated issue, which is where we should put this, plus a link to the actual documentation.

**Review by waynew — DISMISSED** [review]

**Comment by dwoz** [comment]
> ```ini
> [options.entry_points]
> salt.loader=
>   <this-name-does-not-matter = thirpartypackage
>   <this-name-does-not-matter = thirpartypackage:callable
> ```

I think `this-name-does-not-matter` is part of what we want to show up in the versions report, in which case it would matter. The reason being a pypi package might contain more than one 'extension'.

```ini
[options.entrypoints]
salt.loader=
   myplugin1 = myplugin1path
   myplugin2 = myplugin2path
```

At least, I believe this is what was intended for they keys in the configs for entrypoints and can see some value from it.

**Comment by s0undt3ch** [comment]
But then you have the same extension, with the same version, showing up multiple times on the versions report, just because they provide more than one entry point.... I don't see any value added there....

```
third-party-extension[myplugin1,myplugin2,etc...]
```
Would be less verbose, but I still don't see added value there....

**Review by twangboy — DISMISSED** [review]

**Comment by s0undt3ch** [comment]
re-run all

**Comment by dwoz** [comment]
> But then you have the same extension, with the same version, showing up multiple times on the versions report, just because they provide more than one entry point.... I don't see any value added there....
> 
> ```
> third-party-extension[myplugin1,myplugin2,etc...]
> ```
> 
> Would be less verbose, but I still don't see added value there....

It's not about the extension, it's about the plugin. There can be a case where the package providing the plugin has a whole lot of functionality that has nothing to do with Salt. For instance, boto could provide a salt plugin that they maintain which ships with boto3.

**Comment by s0undt3ch** [comment]
> It's not about the extension, it's about the plugin. There can be a case where the package providing the plugin has a whole lot of functionality that has nothing to do with Salt. For instance, boto could provide a salt plugin that they maintain which ships with boto3.
And?

We only report salt related extensions...

I'm not seeing the issue

**Comment by s0undt3ch** [comment]
> For instance, boto could provide a salt plugin that they maintain which ships with boto3.

And we report that boto3 version X.Y.Z provide extensions for salt...

**Review by dwoz — DISMISSED** [review]

File: changelog/58938.added [hunk 2]
```diff
@@ -0,0 +1,24 @@
+Salt's versions report `salt --versions-report` now includes all installed salt extensions into its versions report.
+
+```
+Salt Version:
```
Comment (waynew) [inline changelog/58938.added:4]: **suggestion**: I don't think we want an example here. Changeling entries should be 1 line. They will link back to the issue/pr for more context

**Review by waynew — CHANGES_REQUESTED** [review]
I'll look into this closer tomorrow morning, but in the meantime there is the changelog comment

File: changelog/58938.added (hunk 2 above)
Comment (s0undt3ch) [reply changelog/58938.added]: Resolved

File: salt/loader.py [hunk 3]
```diff
@@ -152,23 +134,93 @@ def _module_dirs(
...
-                try:
+        if ext_type_dirs:
+            for entry_point in entrypoints.iter_entry_points("salt.loader"):
+                with catch_entry_points_exception(entry_point) as ctx:
                     loaded_entry_point = entry_point.load()
-                    for path in loaded_entry_point():
+                if ctx.exception_caught:
+                    continue
+
+                # Old way of defining loader entry points
```
Comment (waynew) [inline salt/loader.py:144]: **suggestion**: mention that this is in setup.cfg. When using `setup.py` you'll want:

```
entry_points = '''
    [salt.loader]
    module_dirs = ...
 ```

File: salt/loader.py [hunk 4]
```diff
@@ -152,23 +134,93 @@ def _module_dirs(
...
+                #   [options.entry_points]
+                #   salt.loader=
+                #     <this-name-does-not-matter> = thirpartypackage
+                #     <this-name-does-not-matter> = thirpartypackage:callable
+                #
+                # We try and see if the thirpartypackage has a `ext_type` sub module, and if so,
+                # we append it to loaded_entry_point_paths.
+                # If the entry-point is in the form of `thirpartypackage:callable`, the return of that
+                # callable must be a dictionary where the keys are the `ext_type`'s and the values must be
+                # lists of paths.
```
Comment (waynew) [inline salt/loader.py:161]: **suggestion**: not sure if we document it elsewhere but we should either mention that here or add some examples of the expected format for this.

File: salt/loader.py [hunk 5]
```diff
@@ -2260,3 +2312,20 @@ def wrapper(*args, **kwargs):
...
+    except Exception as exc:  # pylint: disable=broad-except
+        context.exception_caught = True
+        entry_point_details = entrypoints.name_and_version_from_entry_point(entry_point)
+        log.error(
+            "Error processing Salt Extension %s(version: %s): %s",
+            entry_point_details.name,
+            entry_point_details.version,
+            exc,
+            exc_info_on_loglevel=logging.DEBUG,
+        )
```
Comment (waynew) [inline salt/loader.py:2331]: **question**: Just double checking - this *does* end out dropping the reference to the `exc` stack frames, right? i.e. this isn't introducing any memory leaks.

I'm pretty sure that this is good and we are **not** but I want to make sure I'm not overlooking anything :upside_down_face:

File: tests/support/pytest/helpers.py [hunk 6]
```diff
@@ -533,6 +533,212 @@ def shell_test_false():
...
+            python_requires = >= 3.5
+            setup_requires =
+              wheel
+              setuptools>=50.3.2
+
+            [options.entry_points]
+            salt.loader=
+              module_dirs = {1}
+              runner_dirs = {1}.loader:get_runner_dirs
+              wheel_dirs = {1}.loader:get_new_style_entry_points
```
Comment (waynew) [inline tests/support/pytest/helpers.py:610]: **suggestion**: should also add the negative cases,  e.g. module_dirs` when there's nothing there, or points to the wrong directory, stuff like that.

File: tests/support/pytest/helpers.py [hunk 7]
```diff
@@ -533,6 +533,212 @@ def shell_test_false():
...
+            PKG_ROOT = pathlib.Path(__file__).resolve().parent
+
+            def get_module_dirs():
+                return [str(PKG_ROOT / "modules")]
+
+            def get_runner_dirs():
+                return [str(PKG_ROOT / "runners1"), str(PKG_ROOT / "runners2")]
+
+            def get_new_style_entry_points():
+                return {"wheel": [str(PKG_ROOT / "the_wheel_modules")]}
```
Comment (waynew) [inline tests/support/pytest/helpers.py:635]: **suggestion** similar to above: add some non-existent paths, or paths that have broken imports to ensure that our error reporting is :ok_hand:

**Review by waynew — CHANGES_REQUESTED** [review]
Hey! Sorry this took a bit longer than planned to get back to you. I think this looks excellent!

There are a couple of places where we could add a bit more testing (the negative cases). But otherwise this seems :+1: to me.

I cloned this locally and added my own "extension module" and tried the 3 different styles of loading - list, dict, and `ext_types` sub module + `/modules/` dir and manually verified that it works.

Oh, one more **question**: should we also test the other types of dirs? I see that this test includes modules, but don't we have other potential dirs? from my debugging code:
```
ext_dirs? auth_dirs
ext_dirs? engines_dirs
ext_dirs? fileserver_dirs
ext_dirs? grains_dirs
ext_dirs? log_handlers_dirs
ext_dirs? matchers_dirs
ext_dirs? module_dirs
ext_dirs? returner_dirs
ext_dirs? runner_dirs
ext_dirs? serializers_dirs
ext_dirs? tokens_dirs
ext_dirs? top_dirs
ext_dirs? utils_dirs
ext_dirs? wheel_dirs
```

File: salt/loader.py (hunk 3 above)
Comment (s0undt3ch) [reply salt/loader.py:144]: I think this should be addressed in the upcoming salt extensions documentation, not this particular comment.
If one does not know what this is, one should not mess with it and read docs :)

File: salt/loader.py (hunk 4 above)
Comment (s0undt3ch) [reply salt/loader.py:161]: Yes, in an upcoming documentation only PR which walks you through the process of creating an extension from scratch

**Review by Ch3LL — APPROVED** [review]

File: salt/loader.py (hunk 5 above)
Comment (s0undt3ch) [reply salt/loader.py:2331]: once the log record get's formatted, these references are gone, if they are not, it's the stdlib to blame :grin:

File: tests/support/pytest/helpers.py (hunk 6 above)
Comment (s0undt3ch) [reply tests/support/pytest/helpers.py:610]: Will do on a separate PR

File: tests/support/pytest/helpers.py (hunk 7 above)
Comment (s0undt3ch) [reply tests/support/pytest/helpers.py:635]: Will do on a separate PR

**Review by twangboy — APPROVED** [review]

**Review by krionbsd — APPROVED** [review]
